File: iqGenerator.py
# problem types
# 	sequence
# 		transformations are applied between every item in sequence
# 	combination
# 		middle is combination of the sides

# possible new sequence transformations
# 	char
# 		next alpha
# 		next from sample chars

# ToDo
#  -- Add more problem types

import string
import random
import copy

class RandomChars:

	def __init__(self, countUniqueChars, repeatLimit):
		self.totalCharCount = 0
		self.countUniqueChars = countUniqueChars
		self.repeatLimit = repeatLimit
		self.selectedChars = RandomChars.GetNChars(countUniqueChars)

	def __iter__(self):
		self.totalCharCount = 0
		self.charCounts = {c:0 for c in self.selectedChars}
		return self

	def __next__(self):
		if self.totalCharCount >= self.countUniqueChars * self.repeatLimit:
			raise StopIteration()
		else:
			while True:
				c = random.choice(self.selectedChars)
				if self.charCounts[c] < self.repeatLimit:
					self.totalCharCount += 1
					self.charCounts[c] += 1
					return self.RandomSwapCase(c)

	@staticmethod
	def RandomSwapCase(c):
		if random.random() > .5:
			return c.swapcase()
		return c

	# returns mixed case, but without repeating a letter
	@staticmethod
	def GetNChars(n):
		selectedChars = random.sample(string.ascii_lowercase, n)
		return [c for c in selectedChars]

class CharMatrix:

	def __init__(self, size):
		chars = iter(RandomChars(size+1, size))
		self.size = size
		self.rows = []
		for i in range(size):
			row = []
			for j in range(size):
				row.append(next(chars))
			self.rows.append(row)

	def __iter__(self):
		self._i_ = 0
		return self

	def __next__(self):
		if self._i_ >= self.size**2:
			raise StopIteration()
		result = self.rows[ self._i_//self.size ][ self._i_%self.size ]
		self._i_ += 1
		return result

	def Reflect(self, axis):
		reflected = []
		for j in range(self.size-1, -1, -1):
			reflected.append(self.rows[j])
		self.rows = reflected
		if axis == 'x':
			pass # This if is here just so there is a complete enumeration of possibilities
		elif axis == 'y':
			self.Rotate(180)
		elif axis == 'y=x':
			self.Rotate(270)
		elif axis == 'y=-x':
			self.Rotate(90)
		else:
			raise Exception()

	def Rotate(self, degrees):
		rotated = []
		if degrees == 90 or degrees == -270:
			for j in range(self.size):
				newRow = []
				for i in range(self.size-1, -1, -1):
					newRow.append(self.rows[i][j])
				rotated.append(newRow)
		elif degrees == 180 or degrees == -180:
			for i in range(self.size-1, -1, -1):
				newRow = []
				for j in range(self.size-1, -1, -1):
					newRow.append(self.rows[i][j])
				rotated.append(newRow)
		elif degrees == -90 or degrees == 270:
			for j in range(self.size-1, -1, -1):
				newRow = []
				for i in range(self.size):
					newRow.append(self.rows[i][j])
				rotated.append(newRow)
		else:
			raise Exception()
		self.rows = rotated

	def RowShift(self, row, n):
		n = -n
		self.rows[row] = self.rows[row][n:] + self.rows[row][:n]

	def ColShift(self, col, n):
		tmpCol = []
		for i in range(self.size):
			tmpCol.append(self.rows[i][col])
		for i in range(self.size):
			self.rows[i][col] = tmpCol[(i+n)%self.size]

	def __str__(self):
		horizontalRule = "-" * ( 4*self.size +1 )
		s = horizontalRule + '\n'
		for row in self.rows:
			for c in row:
				s += "| " + c + " "
			s += "|\n" + horizontalRule + '\n'
		return s.rstrip('\n')
		

class Transformation:

	@staticmethod
	def _conditionalDifficulty(oldTrans, newTrans):
		types = set([type(oldTrans), type(newTrans)])
		if SwapCase in types:
			if len(types) == 1 and oldTrans.row == newTrans.row and oldTrans.col == newTrans.col:
				return False
			return newTrans.difficulty
		if Rotate in types:
			if len(types) == 1 or Reflect in types:
				return False
			if HorizontalShift in types or VerticalShift in types:
				return newTrans.difficulty +1
			if RowShift in types or ColShift in types:
				return newTrans.difficulty +2
			raise Exception()
		if Reflect in types:
			if len(types) == 1:
				return False
			if HorizontalShift in types or VerticalShift in types:
				return newTrans.difficulty +1
			if RowShift in types or ColShift in types:
				return newTrans.difficulty +2
			raise Exception()
		if HorizontalShift in types:
			if len(types) == 1:
				return False
			if VerticalShift in types:
				return newTrans.difficulty
			if RowShift in types or ColShift in types:
				return newTrans.difficulty +1
			raise Exception()
		if VerticalShift in types:
			if len(types) == 1:
				return False
			if RowShift in types or ColShift in types:
				return newTrans.difficulty +1
			raise Exception()
		if RowShift in types:
			if len(types) == 1:
				if oldTrans.row == newTrans.row:
					return False
				if abs(oldTrans.row - newTrans.row) == 1 and oldTrans.shift == newTrans.shift:
					return 0.1
				return newTrans.difficulty
			if ColShift in types:
				return newTrans.difficulty +1
			raise Exception()
		if ColShift in types:
			if len(types) == 1:
				if oldTrans.col == newTrans.col:
					return False
				if abs(oldTrans.col - newTrans.col) == 1 and oldTrans.shift == newTrans.shift:
					return 0.1
				return newTrans.difficulty
		raise Exception(str(oldTrans) + ":" + str(newTrans))

	@classmethod
	def ConditionalDifficulty(cls, oldTransfs, newTrans):
		maxDifficulty = newTrans.difficulty
		minDifficulty = newTrans.difficulty
		for trans in oldTransfs:
			conditionalDifficulty = cls._conditionalDifficulty(trans, newTrans)
			if not conditionalDifficulty:
				return False
			if maxDifficulty < conditionalDifficulty:
				maxDifficulty = conditionalDifficulty
			elif minDifficulty > conditionalDifficulty:
				minDifficulty = conditionalDifficulty
		if minDifficulty < newTrans.difficulty:
			return minDifficulty
		return maxDifficulty

	@staticmethod
	def GetRandomTransformation(size):
		transformations = [Rotate, Reflect, HorizontalShift, VerticalShift, RowShift, ColShift, SwapCase]
		ct = len(transformations)
		# the 7 in ct*7 is because the 3 rotations and 4 reflacts are related.
		# To get them with equal frequency, I need to move a 1/7th from rotations to reflections
		weights = [6/(ct*7), 8/(ct*7), 1/ct, 1/ct, 1/ct, 1/ct, 6/ct]
		return random.choices(transformations, weights)[0].GetRandom(size)

	@classmethod
	def GetRandomTransformations(cls, size, targetDifficulty):
		difficulty = 0
		transformations = []
		while difficulty <= targetDifficulty - .5:
			trans = cls.GetRandomTransformation(size)
			trans.difficulty = cls.ConditionalDifficulty(transformations, trans)
			if trans.difficulty and trans.difficulty + difficulty <= targetDifficulty + .5:
				difficulty += trans.difficulty
				transformations.append(trans)
		return transformations

class Rotate(Transformation):

	def __init__(self, degrees):
		self.degrees = degrees
		self.difficulty = 2
		if degrees == 180:
			self.difficulty = 1

	def __str__(self):
		return f"Rotate by {self.degrees} degrees"

	@classmethod
	def GetRandom(cls, size):
		return cls(random.choice((90, 180, 270)))

	def Transform(self, matrix):
		matrix.Rotate(self.degrees)

class Reflect(Transformation):

	def __init__(self, axis):
		self.axis = axis
		self.difficulty = 1

	def __str__(self):
		return f"Matrix was reflected about the {self.axis} axis"

	@classmethod
	def GetRandom(cls, size):
		return cls(random.choice(('x', 'y', 'y=x', 'y=-x')))

	def Transform(self, matrix):
		matrix.Reflect(self.axis)

class HorizontalShift(Transformation):

	def __init__(self, shift):
		self.shift = shift
		self.difficulty = 1

	def __str__(self):
		return f"Matrix was shifted {self.shift} positions horizontally"

	@classmethod
	def GetRandom(cls, size):
		return cls(random.choice(range(1, size)))

	def Transform(self, matrix):
		for i in range(matrix.size):
			matrix.RowShift(i, self.shift)

class VerticalShift(Transformation):

	def __init__(self, shift):
		self.shift = shift
		self.difficulty = 1

	def __str__(self):
		return f"Matrix was shifted {self.shift} positions vertically"

	@classmethod
	def GetRandom(cls, size):
		return cls(random.choice(range(1, size)))

	def Transform(self, matrix):
		for j in range(matrix.size):
			matrix.ColShift(j, self.shift)

class RowShift(Transformation):

	def __init__(self, row, shift):
		self.row = row
		self.shift = shift
		self.difficulty = 2
		if abs(shift) == 1:
			self.difficulty = 1

	def __str__(self):
		return f"Row {self.row} was shifted {self.shift} positions horizontally (zero indexing)"

	@classmethod
	def GetRandom(cls, size):
		return cls(random.choice(range(size)), random.choice(range(1, size)))

	def Transform(self, matrix):
		matrix.RowShift(self.row, self.shift)

class ColShift(Transformation):

	def __init__(self, col, shift):
		self.col = col
		self.shift = shift
		self.difficulty = 2
		if abs(shift) == 1:
			self.difficulty = 1

	def __str__(self):
		return f"Column {self.col} was shifted {self.shift} positions vertically (zero indexing)"

	@classmethod
	def GetRandom(cls, size):
		return cls(random.choice(range(size)), random.choice(range(1, size)))

	def Transform(self, matrix):
		matrix.ColShift(self.col, self.shift)

class SwapCase(Transformation):

	def __init__(self, row, col):
		self.row = row
		self.col = col
		self.difficulty = 0.25

	def __str__(self):
		return f"The case of row {self.row} column {self.col} has been swapped (zero indexing)"

	@classmethod
	def GetRandom(cls, size):
		return cls(random.choice(range(size)), random.choice(range(size)))

	def Transform(self, matrix):
		matrix.rows[self.row][self.col] = matrix.rows[self.row][self.col].swapcase()


class Sequence:

	def __init__(self, size, difficulty):
		self.size = size
		self.matricies = []
		self.matricies.append(CharMatrix(size))
		self.transformations = Transformation.GetRandomTransformations(self.size, difficulty)
		for m in range(3):
			nextMatrix = copy.deepcopy(self.matricies[-1])
			for trans in self.transformations:
				trans.Transform(nextMatrix)
			self.matricies.append(nextMatrix)

	def _matrixRowToStr(self, start, stop):
		horizontalRule = ("-" * ( 4*self.size +1 ) + "  ") * (stop - start)
		s = horizontalRule + "\n"
		for i in range(self.size):
			for matrix in self.matricies[start:stop]:
				for c in matrix.rows[i]:
					s += "| " + c + " "
				s += "|  "
			s += "\n" + horizontalRule + "\n"
		return s

	def DisplayProblem(self):
		s = ""
		s += self._matrixRowToStr(0, 3) + "\n"
		print(s)

	def DisplaySolution(self):
		print("\n" + str(self.matricies[-1]) + "\n")
		for transformation in self.transformations:
			print(transformation)
		print()


def GetMatrixSize():
	while True:
		matrixSize = input("Enter matrix size (4): ")
		if not matrixSize:
			matrixSize = 4
		try:
			matrixSize = int(matrixSize)
			if matrixSize < 3:
				print("Matrix size must be >=3")
			else:
				return matrixSize
		except Exception:
			print("Matrix size must be an integer >=3.")

def GetDifficulty():
	while True:
		difficultyStr = input("Easy (e), Medium (m) (default), or Hard (h)?: ")
		if not difficultyStr:
			difficultyStr = 'm'
		difficultyStr = difficultyStr.lower()
		if difficultyStr == 'easy' or difficultyStr == 'e':
			return 2.5
		if difficultyStr == 'medium' or difficultyStr == 'm':
			return 4
		if difficultyStr == 'hard' or difficultyStr == 'h':
			return 6
		print("Please select a valid difficulty.")

instructions = '''Instructions:
   The leftmost matrix is the start matrix.
   The center matrix is produced by applying a series of transformations to the left matrix.
   The right matrix is produced by applying the same transformations to the center matrix.
   If we apply the same transformations to the right matrix, we will get the final matrix.
   Try to predict the final matrix.'''

def main():

	print()
	matrixSize = GetMatrixSize()
	print()
	difficulty = GetDifficulty()
	print()
	print(instructions)
	print()

	while  True:
		seq = Sequence(matrixSize, difficulty)
		seq.DisplayProblem()
		input("Press enter for solution: ")
		seq.DisplaySolution()
		input("\n Press enter for new problem: ")
		print("\n  ---------  new problem  ----------  \n")


main()

File: iqGenerator.hs
import Data.List (transpose, intercalate, intersperse)
import Data.Char (toLower, toUpper, isUpper)
import System.Random.Stateful (StdGen, uniformR, uniform, mkStdGen)
import Data.Maybe (isNothing)
import Text.Read (readMaybe)
import GHC.IO.Handle (hFlush)
import GHC.IO.Handle.FD (stdout)
import Data.Time.Clock.System (getSystemTime, SystemTime (systemSeconds))
import Control.Monad.Trans.State ( State, state, runState )
import Control.Monad (replicateM)
import Control.Applicative (Applicative(liftA2))

main :: IO ()
main = do
    mSize <- getMatrixSize
    d <- difficultyScaler <$> getDifficulty
    sysTime <- getSystemTime
    let secs = fromIntegral $ systemSeconds sysTime
    let g = mkStdGen secs

    let (m1, g') = runState (randMatrix mSize) g
    let (ts, _) = runState (genTransforms mSize d) g'
    let t = foldl1 (.) (getTransform <$> ts)

    let m2 = t m1
    let m3 = t m2

    putLn
    print m1
    putLn
    print m2
    putLn
    putStr' "Press Enter to see solution: "
    _ <- getLine
    putLn
    putStrLn $ "Using transforms: " ++ show ts
    putLn
    print m3
    putLn


-- ### Data Structures ###

type Row = [Char]

newtype Matrix = ToM { froM :: [Row] }
instance Show Matrix
    where show (ToM rows) = ' ' : intercalate "\n " [intersperse ' ' r | r <- rows]

data Rotation = R90 | R180 | R270 deriving (Eq, Show, Enum)

data Axis = X | Y | XY | NXY deriving (Eq, Show, Bounded, Enum)

data Transform =
    Rotate Rotation
    | Reflect Axis
    | HorizontalShift Int
    | VerticalShift Int
    | RowShift Int Int
    | ColShift Int Int
    | SwapCase Int Int
    deriving (Eq, Show)

data Difficulty = Easy | Medium | Hard


-- ### Matrix Transformations ###

-- | Transform datum to transforming function
getTransform :: Transform -> (Matrix -> Matrix)
getTransform t = case t of
    Rotate R90        -> rotate90
    Rotate R180       -> rotate180
    Rotate R270       -> rotate270
    Reflect X         -> xReflect
    Reflect Y         -> yReflect
    Reflect XY        -> xyReflect
    Reflect NXY       -> nxyReflect
    HorizontalShift s -> hShift s
    VerticalShift s   -> vShift s
    RowShift row s    -> rShift row s
    ColShift col s    -> cShift col s
    SwapCase row col  -> swapCase' row col

-- | Perform row operation op on one based index row r of matrix
rowOp :: Int -> (Row -> Row) -> Matrix -> Matrix
rowOp r op = ToM . elemOp r op . froM

-- | Perform "row" operation on one based index column c of matrix
colOp :: Int -> (Row -> Row) -> Matrix -> Matrix
colOp c op = nxyReflect . rowOp c op . nxyReflect

-- | Perform cell operation on one based indecies row r column c of matrix
cellOp :: Int -> Int -> (Char -> Char) -> Matrix -> Matrix
cellOp r c op = rowOp r (elemOp c op)

-- | Shift matrix n places to the left. Use negative n for right shifts
hShift :: Int -> Matrix -> Matrix
hShift n = ToM . map (shift n) . froM

-- | Shift matrix n places up. Use negative n to shift down
vShift :: Int -> Matrix -> Matrix
vShift n = ToM . shift n . froM

-- | Shift row r s places to the left
rShift :: Int -> Int -> Matrix -> Matrix
rShift r s = rowOp r (shift s)

-- | Shift column c s places up
cShift :: Int -> Int -> Matrix -> Matrix
cShift c s = colOp c (shift s)

-- | Reflect matrix over the x-axis
xReflect :: Matrix -> Matrix
xReflect = ToM . reverse . froM

-- | Reflect matrix over the y-axis
yReflect :: Matrix -> Matrix
yReflect = ToM . map reverse . froM

-- | Reflect matrix over the y=-x axis
nxyReflect :: Matrix -> Matrix
nxyReflect = ToM . transpose . froM

-- | Reflect matrix over the y=x axis
xyReflect :: Matrix -> Matrix
xyReflect = xReflect . nxyReflect . xReflect

rotate180 :: Matrix -> Matrix
rotate180 = xReflect . yReflect

rotate90 :: Matrix -> Matrix
rotate90 = nxyReflect . xReflect

rotate270 :: Matrix -> Matrix
rotate270 = xReflect . nxyReflect

-- | Swap the case of the char at row r column c
swapCase' :: Int -> Int -> Matrix -> Matrix
swapCase' r c = cellOp r c swapCase


-- ### Difficulty Calculation ###

-- | Difficulty setting to difficulty number
difficultyScaler :: Difficulty -> Float
difficultyScaler Easy = 2.5
difficultyScaler Medium = 4
difficultyScaler Hard = 6

-- | Independent difficulty of a transform t
difficulty :: Transform -> Float
difficulty t = case t of
    Rotate degree     -> if degree == R180 then 1 else 2
    Reflect _         -> 1
    HorizontalShift _ -> 1
    VerticalShift _   -> 1
    RowShift _ s      -> if abs s == 1 then 1 else 2
    ColShift _ s      -> if abs s == 1 then 1 else 2
    SwapCase _ _      -> 0.25

-- | Difficulty of all the transforms together
--  If any transform is incompatible with any other, we get Nothing
cumulativeDifficulty :: [Transform] -> Maybe Float
cumulativeDifficulty [] = Just 0
cumulativeDifficulty [t] = Just $ difficulty t
cumulativeDifficulty (t:ts) = (+)
    <$> additionalDifficulty t ts
    <*> cumulativeDifficulty ts

-- | Calculate the marginal difficulty of adding transform t to a set of existing transforms ts
--  Return is maybe because some transforms are incompatible. If t is incompatible with any of ts, we get Nothing
additionalDifficulty :: Transform -> [Transform] -> Maybe Float
additionalDifficulty t [] = Just $ difficulty t
additionalDifficulty t ts =
    if tDifficulty > minDifficulty
        then minDifficulty
        else maxDifficulty
    where condDifficulties = fmap (conditionalDifficulty t) ts
          maxDifficulty = foldl1 maxFail condDifficulties
          minDifficulty = foldl1 minFail condDifficulties
          tDifficulty = Just $ difficulty t

-- | Calculate the marginal difficulty of adding a transform newT given there is already a transform oldT
--  Return is maybe because some transforms are incompatible, in which case we get Nothing
conditionalDifficulty :: Transform -> Transform -> Maybe Float
conditionalDifficulty newT oldT
    | oldT == newT             = Nothing
    | shiftsOverlap oldT newT  = Nothing
    | shiftsCoincide oldT newT = Just 0.1
    | otherwise                = case (newT, oldT) of

        -- SwapCase.
        -- Swapping the case of a cell does not interact significantly with other transformations
        (SwapCase{}, _)                        -> Just newDifficulty
        (_, SwapCase{})                        -> Just newDifficulty

        -- Rotate
        (Rotate{}, Rotate{})                   -> Nothing
        -- Combining a rotation and a reflection results in a different reflection, so we do not combine them 
        (Rotate{}, Reflect{})                  -> Nothing
        (Reflect{}, Rotate{})                  -> Nothing
        (Rotate{}, HorizontalShift{})          -> Just $ newDifficulty + 1.5
        (HorizontalShift{}, Rotate{})          -> Just $ newDifficulty + 1.5
        (Rotate{}, VerticalShift{})            -> Just $ newDifficulty + 1.5
        (VerticalShift{}, Rotate{})            -> Just $ newDifficulty + 1.5
        -- I do not recall why I decided rotation combined with row or column shifts is harder. Note, the same for Reflect below
        (Rotate{}, _)                          -> Just $ newDifficulty + 2
        (_, Rotate{})                          -> Just $ newDifficulty + 2

        -- Reflect
        (Reflect{}, Reflect{})                 -> Nothing
        (Reflect{}, HorizontalShift{})         -> Just $ newDifficulty + 1.5
        (HorizontalShift{}, Reflect{})         -> Just $ newDifficulty + 1.5
        (Reflect{}, VerticalShift{})           -> Just $ newDifficulty + 1.5
        (VerticalShift{}, Reflect{})           -> Just $ newDifficulty + 1.5
        (Reflect{}, _)                         -> Just $ newDifficulty + 2
        (_, Reflect{})                         -> Just $ newDifficulty + 2

        -- HorizontalShift
        (HorizontalShift{}, HorizontalShift{}) -> Nothing
        (HorizontalShift{}, VerticalShift{})   -> Just newDifficulty
        (VerticalShift{}, HorizontalShift{})   -> Just newDifficulty
        (HorizontalShift{}, _)                 -> Just $ newDifficulty + 1
        (_, HorizontalShift{})                 -> Just $ newDifficulty + 1

        -- VerticalShift
        (VerticalShift{}, VerticalShift{})     -> Nothing
        (VerticalShift{}, _)                   -> Just $ newDifficulty + 1
        (_, VerticalShift{})                   -> Just $ newDifficulty + 1

        -- RowShift
        (RowShift{}, _)                        -> Just $ newDifficulty + 1
        (_, RowShift{})                        -> Just $ newDifficulty + 1

        -- ColShift
        (ColShift{}, _)                        -> Just $ newDifficulty + 1

    where
        newDifficulty = difficulty newT
        shiftsOverlap t1 t2 = case (t1, t2) of
            (HorizontalShift{}, HorizontalShift{}) -> True
            (VerticalShift{}, VerticalShift{})     -> True
            (RowShift r1 _, RowShift r2 _)         -> r1 == r2
            (ColShift c1 _, ColShift c2 _)         -> c1 == c2
            (_, _)                                 -> False
        shiftsCoincide t1 t2 = case (t1, t2) of
            (RowShift r1 s1, RowShift r2 s2) -> abs (r1 - r2) == 1 && s1 == s2
            (ColShift c1 s1, ColShift c2 s2) -> abs (c1 - c2) == 1 && s1 == s2
            (_, _)                           -> False


-- ### Random Board and Transform Generation ###

genTransforms :: Int -> Float -> State StdGen [Transform]
genTransforms mSize targetD = genTransforms' mSize targetD []

genTransforms' :: Int -> Float -> [Transform] -> State StdGen [Transform]
genTransforms' mSize targetDifficulty priorTs = do
    generator <- selectRandom transformGenerators
    newT <- generator mSize
    let ts = newT:priorTs
    let d = cumulativeDifficulty ts
    let overshot = d > Just (targetDifficulty + 0.5)
    let reached = (abs . (-) targetDifficulty <$> d) <= Just 0.5
    if isNothing d || overshot then
        genTransforms' mSize targetDifficulty priorTs
    else if reached then
        return ts
    else
        genTransforms' mSize targetDifficulty ts

transformGenerators :: [Int -> State StdGen Transform]
transformGenerators = 
    [\_ -> do
        r <- selectRandom [R90, R180, R270]
        return $ Rotate r
    ,\_ -> do
        a <- selectRandom [X,Y, XY, NXY]
        return $ Reflect a
    ,\mSize -> do
        s <- state $ uniformR (1, mSize - 1)
        return $ HorizontalShift s
    ,\mSize -> do
        s <- state $ uniformR (1, mSize - 1)
        return $ VerticalShift s
    ,\mSize -> do
        row <- state $ uniformR (1, mSize)
        s <- state $ uniformR (1, mSize - 1)
        return $ RowShift row s
    ,\mSize -> do
        col <- state $ uniformR (1, mSize)
        s <- state $ uniformR (1, mSize - 1)
        return $ ColShift col s
    ,\mSize -> do
        row <- state $ uniformR (1, mSize)
        col <- state $ uniformR (1, mSize)
        return $ SwapCase row col]

-- | Returns a randomly selected item from a list. Beware of passing the empty list
selectRandom :: [a] -> State StdGen a
selectRandom xs = do
    let l = length xs
    i <- state $ uniformR (0, l - 1)
    return $ xs !! i

-- | Returns a randomly generated square matrix of size n
randMatrix :: Int -> State StdGen Matrix
randMatrix n = do
    rows <- replicateM n (randRow n)
    return $ ToM rows

-- | Retruns a randomly generated string of length n
randRow ::  Int -> State StdGen [Char]
randRow n = replicateM n randChar

-- | Returns a randomly selected letter from A-Z or a-z
randChar :: State StdGen Char
randChar = do
    c <- state $ uniformR ('a', 'z')
    randSwapCase c

-- | Randomizes the case of c
randSwapCase :: Char -> State StdGen Char
randSwapCase c = do
    condSwap c <$> randBool

randBool :: State StdGen Bool
randBool = state uniform


-- ### CLI IO ###

-- | Prompt the user to enter a matrix size. Must be 3 or larger
getMatrixSize :: IO Int
getMatrixSize = do
    putStr' "Enter Matrix Size >= 3: "
    sizeString <- getLine
    let maybeSize = readMaybe sizeString :: Maybe Int
    mSize <- maybe getMatrixSize return maybeSize
    if mSize >= 3
        then return mSize
        else getMatrixSize

-- | Prompt the user to enter a difficulty of easy, medium, or hard
getDifficulty :: IO Difficulty
getDifficulty = do
    putStr' "Easy (e) (default), Medium (m), or Hard (h)?: "
    dStr <- getLine
    let d = parseDifficulty dStr
    maybe getDifficulty return d

parseDifficulty :: String -> Maybe Difficulty
parseDifficulty d
    | d' == "e" || d' == "easy" || d' == "" = Just Easy
    | d' == "m" || d' == "medium" = Just Medium
    | d' == "h" || d' == "hard" = Just Hard
    | otherwise = Nothing
    where d' = toLower <$> d

-- | Prints the string and forces stdout flush
putStr' :: String -> IO ()
putStr' s = do
    putStr s
    hFlush stdout

-- | Prints a newline
putLn :: IO ()
putLn = putStrLn ""


-- ### Helper Functions ###

-- | If b, the swap the case of char c. Otherwise, return c
condSwap :: Char -> Bool -> Char
condSwap c b = if b then swapCase c else c

swapCase :: Char -> Char
swapCase c
    | isUpper c = toLower c
    | otherwise = toUpper c

-- ### Shift list xs n places to the left
shift :: Int -> [a] -> [a]
shift n xs = drop n xs ++ take n xs

-- | Peform operation op on element at one based index op
elemOp :: Int -> (a -> a) -> [a] -> [a]
elemOp idx op xs = start ++ op nth : end
    where start = take (idx-1) xs
          nth = xs !! (idx-1)
          end = drop idx xs

-- | If both inputs are Just, return max. Otherwise, return nothing
maxFail :: (Ord a) => Maybe a -> Maybe a -> Maybe a
maxFail = liftA2 max

-- | If both inputs are Just, return min. Otherwise, return nothing
minFail :: (Ord a) => Maybe a -> Maybe a -> Maybe a
minFail = liftA2 min